quickcache/quickcache.py

```diff
@@ -1,329 +1,67 @@
-
-          now as soon as request.couch_user._rev has changed,
-          the function will be recomputed
-
-    Features:
-        - In addition to caching in the default shared cache,
-          quickcache caches in memory for 10 seconds
-          (conceptually the length of a single request).
-          This can be overridden to a different number with memoize_timeout.
-
-        - In addition to varying on the arguments and the name of the function,
-          quickcache will also make sure to vary
-          on the _source code_ of your function.
-          That way if you change the behavior of the function, there won't be
-          any stale cache when you deploy.
-
-        - Can vary on any number of the function's parameters
-
-        - Does not by default vary on all function parameters.
-          This is because it is not in general obvious what it means
-          to vary on an object, for example.
-
-        - Allows you to vary on an attribute of an object,
-          multiple attrs of the same object, attrs of attrs of an object, etc
-
-        - Allows you to pass in a function as the vary_on arg which will get called
-          with the same args and kwargs as the function. It should return a list of simple
-          values to be used for generating the cache key.
-
-        Note on unicode and strings in vary_on:
-          When strings and unicode values are used as vary on parameters they will result in the
-          same cache key if and only if the string values are UTF-8 or ascii encoded.
-          e.g.
-          u'namé' and 'nam\xc3\xa9' (UTF-8 encoding) will result in the same cache key
-          BUT
-          u'namé' and 'nam\xe9' (latin-1 encoding) will NOT result in the same cache key
-
-
-    """
-    if cache and timeout:
-        raise ValueError('You can only use timeout '
-                         'when not overriding the cache')
-
-    timeout = timeout if timeout is not None else 5 * 60
-    memoize_timeout = memoize_timeout or 10
-
-    if cache is None:
-        caches = [
-            CacheWithTimeout(
-                django_caches['locmem'],
-                timeout=memoize_timeout)]
-        if timeout > 0:
-            caches.append(
-                CacheWithTimeout(
-                    django_caches['default'], timeout=timeout))
-        cache = TieredCache(caches)
-
-    def decorator(fn):
-        helper = helper_class(fn, vary_on=vary_on, cache=cache)
-
-        @functools.wraps(fn)
-        def inner(*args, **kwargs):
-            return helper(*args, **kwargs)
-
-        inner.clear = helper.clear
-        inner.get_cache_key = helper.get_cache_key
-        inner.prefix = helper.prefix
-        inner.get_cached_value = helper.get_cached_value
-
-        return inner
-
-    return decorator
+get_quickcache = QuickCache(
+    vary_on=Ellipsis,
+    cache=Ellipsis,
+    skip_arg=None,
+    helper_class=QuickCacheHelper,
+    assert_function=assert_function,
+).but_with
```

Comment:
I chose `but_with` so that `quickcache.but_with(timeout=60 * 60)` reads out loud like "quickcache, but with timeout set to 60 minutes", which specifically makes it clear that (1) a new object is returned and (2) your original `quickcache` is left untouched. Alternatives like `replace`, `override`, `config` (which is what I'd originally gone with), make it sound like you're doing something to the object. Maybe you still think it sounds weird, maybe because "but with" isn't a verb phrase, and methods are usually verb phrases. I'm open to changing it, but I still personally like `but_with`. On the other hand, if you think it feels weird, other people probably also think it feels weird.